### PR TITLE
pass ModelConfigInfo down to client.Get() for self-hosted-models

### PR DESCRIPTION
**Pull request by slimsag**
This PR has a single goal: pass a new `ModelConfigInfo` type, which has the `Provider` and `Model` we should use to serve a completions request, down into `client.Get()`

This PR explicitly only handles the case we care about for Self-hosted models, and all pieces of logic that I expect will be replaced/superseded by your work @chrsmith are annotated with `// TODO(slimsag): self-hosted-models:` comments so we can easily find and remove them when your work is ready.

Every location I have modified has been carefully wrapped in an if statement like `if conf.Get().SiteConfig().ModelConfiguration != nil` to ensure that this change _only_ affects people who set the new `"modelConfiguration"` site config property.

## Test plan

1. Configured `"modelConfiguration"` and removed `"completions"` in my dev instances' site config.
2. Used VS Code chat to confirm I see this codepath is connected end-to-end and I get a `TODO` error back:

image

## Changelog

N/A

### CODE REVIEW

File: internal/completions/client/client.go [hunk 1]
```diff
@@ -24,16 +25,25 @@ func Get(
...
 
-func getBasic(endpoint string, provider conftypes.CompletionsProviderName, accessToken string) (types.CompletionsClient, error) {
+func getBasic(endpoint string, provider conftypes.CompletionsProviderName, accessToken string, modelConfigInfo *types.ModelConfigInfo) (types.CompletionsClient, error) {
 	tokenManager := tokenusage.NewManager()
+
+	if conf.Get().SiteConfig().ModelConfiguration != nil {
+		// Using the new "modelConfiguration" site config
+		// TODO(slimsag): self-hosted-models: this logic only handles Cody Enterprise with Self-hosted models
+		// Only in this case do we have modelConfigInfo != nil
+		return nil, errors.Newf("TODO: implement self-hosted-models")
```
Comment (slimsag) [inline internal/completions/client/client.go:44]: This is all in service of having `modelConfigInfo` right here.

File: cmd/frontend/internal/httpapi/completions/get_model.go [hunk 2]
```diff
@@ -23,6 +27,21 @@ type getModelFn func(ctx context.Context, requestParams types.CodyCompletionRequ
 
 func getCodeCompletionModelFn() getModelFn {
 	return func(_ context.Context, requestParams types.CodyCompletionRequestParameters, c *conftypes.CompletionsConfig) (string, error) {
+		if modelConfig := conf.Get().SiteConfig().ModelConfiguration; modelConfig != nil {
+			// Using new "modelConfiguration" site config.
+			// TODO(slimsag): self-hosted-models: currently this logic only handles Cody Enterprise with Self-hosted models
+			if err := modelconfig.ValidateModelRef(modelconfigtypes.ModelRef(requestParams.Model)); err != nil {
```
Comment (chrsmith) [inline cmd/frontend/internal/httpapi/completions/get_model.go:33]: This code seems a little hard to follow since you're essentially adding fallback options in case there is an error. So what do you think about inverting the error condition so this is a little easier to follow, e.g.:

```go
// WDYT of using 'modelconfigSDK' rather than 'modelconfigtypes'?
import modelconfigSDK ".../types"

// Check if the user supplied a valid ModelRef.
if requestParams.Model != "" {
    mref := modelconfigSDK.ModelRef(requestParams.Model)
    if err := modelconfig.ValidateModelRef(mref); err == nil {
        return requestParams.Model, nil
    }
}

// Fallback to using the default models as per the configuration.
modelConfig, err := frontendmodelconfig.Get().Get()
if err != nil {
    return "", err
}
return string(modelConfig.DefaultModels.CodeCompletion), nil
```

File: cmd/frontend/internal/httpapi/completions/get_model.go [hunk 3]
```diff
@@ -42,6 +61,24 @@ func getCodeCompletionModelFn() getModelFn {
 
 func getChatModelFn(db database.DB) getModelFn {
 	return func(ctx context.Context, requestParams types.CodyCompletionRequestParameters, c *conftypes.CompletionsConfig) (string, error) {
+		if modelConfig := conf.Get().SiteConfig().ModelConfiguration; modelConfig != nil {
```
Comment (chrsmith) [inline cmd/frontend/internal/httpapi/completions/get_model.go:64]: Since we are using this same stanza in 3x places to mean the same thing, what do you think about creating a one-off function to be clearer about the intent? e.g.

```go
func useExperimentalModelConfigurationSettings() bool
```

File: cmd/frontend/internal/httpapi/completions/handler.go [hunk 4]
```diff
@@ -176,12 +180,29 @@ func newCompletionsHandler(
 			}
 		}
 
+		var modelConfigInfo *types.ModelConfigInfo
+		if conf.Get().SiteConfig().ModelConfiguration != nil {
+			// TODO(slimsag): self-hosted-models: this logic only handles Cody Enterprise with Self-hosted models
+			modelConfig, err := modelconfig.Get().Get()
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
```
Comment (chrsmith) [inline cmd/frontend/internal/httpapi/completions/handler.go:189]: For both of these 500 return paths, we use `logger.Error(...)`. Since otherwise it'll be a pain to debug or troubleshoot any issues users run into.

**Review by chrsmith — APPROVED** [review]
I have suggestions for how we can rewrite the logic within `getModel` to be clearer. But given that I'll rewrite the function entirely in XX hours, I wouldn't worry too much about it. :D

File: cmd/frontend/internal/httpapi/completions/get_model.go (hunk 2 above)
Comment (slimsag) [reply cmd/frontend/internal/httpapi/completions/get_model.go]: 👍 on `modelconfigSDK`
